core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -0,0 +1,131 @@
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadContext);
+        final ClusterState observedState = observer.observedState();
+        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+            @Override
+            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                return newStatus == ClusterState.ClusterStateStatus.APPLIED
+                           && enoughShardsActive(newState, indexName, settings, waitOnActiveShards);
+            }
+
+            @Override
+            public boolean apply(ClusterChangedEvent changedEvent) {
+                return enoughShardsActive(changedEvent.state(), indexName, settings, waitOnActiveShards);
+            }
+        };
+        final ClusterStateObserver.Listener observerListener = new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                listener.onResponse(new ClusterStateUpdateResponse(true));
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
+                                "to be allocated on index creation.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
+                                "number of shards to be allocated.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
+            }
+        };
+
+        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
+                && enoughShardsActive(observedState, indexName, settings, waitOnActiveShards)) {
+            observerListener.onNewClusterState(observedState);
+        } else {
+            observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
+        }
+    }
+
+    // package private for testing
+    static boolean enoughShardsActive(final ClusterState clusterState, final String indexName,
+                                      final Settings settings, int waitOnActiveShards) {
+        if (waitOnActiveShards == ReplicationOperation.WAIT_ON_ACTIVE_SHARDS_NOT_SET) {
+            waitOnActiveShards = WAIT_ON_ACTIVE_SHARDS_SETTING.get(settings);
+        }
+        final IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(indexName);
+        assert indexRoutingTable != null;
+        if (indexRoutingTable.allPrimaryShardsActive() == false) {
```

Comment:
what if I don't want to require for any shards to be available after index creation?